cpp/src/reshape/explode.cu

```diff
@@ -39,77 +39,114 @@ namespace {
+    auto sliced_child = lc.get_sliced_child(stream);
+    rmm::device_uvector<size_type> gather_map_indices(sliced_child.size(), stream, mr);
+
+    // sliced columns can make this a little tricky. We have to start iterating at the start of the
+    // offsets for this column, which could be > 0. Then we also have to handle rebasing the offsets
+    // as we go.
+    auto offsets           = lc.offsets().begin<size_type>() + lc.offset();
+    auto offsets_minus_one = thrust::make_transform_iterator(
+      offsets, [offsets] __device__(auto i) { return (i - offsets[0]) - 1; });
+    auto counting_iter = thrust::make_counting_iterator(0);
+
+    rmm::device_uvector<size_type> pos(include_pos ? sliced_child.size() : 0, stream, mr);
+
+    // This looks like an off-by-one bug, but what is going on here is that we need to reduce each
+    // result from `lower_bound` by 1 to build the correct gather map. It was pointed out that
+    // this can be accomplished by simply skipping the first entry and using the result of
+    // `lower_bound` directly.
+    if (include_pos) {
+      thrust::transform(rmm::exec_policy(stream),
+                        counting_iter,
+                        counting_iter + gather_map_indices.size(),
+                        gather_map_indices.begin(),
+                        [position_array = pos.data(),
+                         offsets_minus_one,
+                         offsets,
+                         offset_size = lc.size()] __device__(auto idx) -> size_type {
+                          auto lb_idx = thrust::lower_bound(thrust::seq,
+                                                            offsets_minus_one + 1,
+                                                            offsets_minus_one + offset_size + 1,
+                                                            idx) -
+                                        (offsets_minus_one + 1);
+                          position_array[idx] = idx - (offsets[lb_idx] - offsets[0]);
+                          return lb_idx;
+                        });
+    } else {
+      thrust::lower_bound(rmm::exec_policy(stream),
+                          offsets_minus_one + 1,
+                          offsets_minus_one + lc.size() + 1,
+                          counting_iter,
+                          counting_iter + gather_map_indices.size(),
+                          gather_map_indices.begin());
+    }
+
+    auto select_iter = thrust::make_transform_iterator(
+      thrust::make_counting_iterator(0),
+      [explode_column_idx](size_type i) { return i >= explode_column_idx ? i + 1 : i; });
+    std::vector<size_type> selected_columns(select_iter,
+                                            select_iter + input_table.num_columns() - 1);
+
+    auto gathered_table = cudf::detail::gather(
+      input_table.select(selected_columns),
+      column_view(
+        data_type(type_to_id<size_type>()), sliced_child.size(), gather_map_indices.data()),
```

Comment:
Can you just use the `gather` variant that takes an iterator? Any time you're having to manually construct a `column_view` to call another function should be a indicator that there may be a better way to do it.